pkg/algorithm/nodeselect/v1alpha2/select_node.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2018 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	bd "github.com/openebs/maya/pkg/blockdevice/v1alpha2"
+	bdc "github.com/openebs/maya/pkg/blockdeviceclaim/v1alpha1"
+	csp "github.com/openebs/maya/pkg/cstor/newpool/v1alpha3"
+	"github.com/openebs/maya/pkg/volume"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GetCandidateNodeMap returns a map of all nodes where the pool needs to be created.
+func (ac *Config) GetCandidateNodeMap() (map[string]bool, error) {
+	// TODO : Do not select a node if it is not ready
+	candidateNodesMap := make(map[string]bool)
+	usedNodeMap, err := ac.GetUsedNodeMap()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get candidate nodes for pool creation")
+	}
+	for _, pool := range ac.CSPC.Spec.Pools {
+		nodeName := pool.NodeSelector[HostName]
+		if usedNodeMap[nodeName] == false {
+			candidateNodesMap[nodeName] = true
+		}
+	}
+	return candidateNodesMap, nil
+}
+
+// GetUsedNodeMap returns a map of node for which pool has already been created.
+func (ac *Config) GetUsedNodeMap() (map[string]bool, error) {
+	usedNodeMap := make(map[string]bool)
+	cspList, err := csp.
+		NewKubeClient().
+		WithNamespace(ac.Namespace).
+		List(
+			metav1.
+				ListOptions{LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + ac.CSPC.Name},
+		)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list already created csp(s)")
+	}
+	for _, cspObj := range cspList.Items {
+		usedNodeMap[cspObj.Labels[string(apis.HostNameCPK)]] = true
+	}
+	return usedNodeMap, nil
+}
+
+// SelectNode selects a node and returns the pool spec from the cspc for pool provisioning.
+func (ac *Config) SelectNode() (*apis.PoolSpec, error) {
+	candidateNodes, err := ac.GetCandidateNodeMap()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get pool spec for pool creation")
+	}
+	for _, pool := range ac.CSPC.Spec.Pools {
+		pool := pool
+		nodeName := pool.NodeSelector[HostName]
+		if candidateNodes[nodeName] {
+			if ValidatePoolSpec(&pool) {
```

Comment:
`ValidatePoolSpec` returning error should makes sense, and returning all those errors at the end of for loop